Sources/SourceKitLSP/Swift/CodeActions/ConvertStringConcatenationToStringInterpolation.swift

```diff
@@ -0,0 +1,139 @@
+
+struct ConvertStringConcatenationToStringInterpolation: SyntaxRefactoringProvider {
+  static func refactor(syntax: SequenceExprSyntax, in context: Void) -> SequenceExprSyntax? {
+    guard let (componentsOnly, commonPounds) = preflight(exprList: syntax.elements) else {
+      return nil
+    }
+
+    var ret: StringLiteralSegmentListSyntax = []
+    for component in componentsOnly {
+      if let stringLiteral = StringLiteralExprSyntax(component) {
+        var segments = stringLiteral.segments
+        if stringLiteral.openingPounds?.tokenKind != commonPounds?.tokenKind {
+          for i in segments.indices {
+            guard case var .expressionSegment(exprSegment) = segments[i] else {
+              continue
+            }
+            exprSegment.pounds = commonPounds
+            segments[i] = .expressionSegment(exprSegment)
+          }
+        }
+        ret += segments
+      } else {
+        ret.append(
+          .expressionSegment(
+            ExpressionSegmentSyntax(
+              pounds: commonPounds,
+              expressions: [
+                LabeledExprSyntax(expression: component.trimmed)
+              ]
+            )
+          )
+        )
+      }
+    }
+
+    return SequenceExprSyntax(elements: [
+      ExprSyntax(
+        StringLiteralExprSyntax(
+          openingPounds: commonPounds,
+          openingQuote: "\"",
+          segments: ret,
+          closingQuote: "\"",
+          closingPounds: commonPounds
+        )
+      )
+    ])
+  }
+
+  /// If `exprList` is a string concatenation, returns 1) all elements in `exprList` with concat operators stripped and 2) the longest pounds amongst all string literals,
+  /// otherwise returns nil.
+  ///
+  ///  `exprList` as a valid string concatenation must contain n >= 3 children where n is an odd number with a concat operator `+` separating every other child which must either be a string literal or a valid expression for string interpolation. `exprList` must also contain at least one string literal child.
+  ///
+  ///  For example,
+  ///
+  ///      "Hello " + aString + "\(1)World"
+  ///  is a valid string concatenation.
+  ///
+  ///      aString + bString
+  ///  and
+  ///
+  ///      "Hello " * aString - "World"
+  ///  are invalid string concatenations.
+  static func preflight(
+    exprList: ExprListSyntax
+  ) -> (componentsOnly: [ExprListSyntax.Element], longestPounds: TokenSyntax?)? {
+    var iter = exprList.makeIterator()
+    guard let first = iter.next() else {
+      return nil
+    }
+
+    var hasStringLiterals = false
+    var longestPounds: TokenSyntax?
+    var componentsOnly = [first]
+
+    if let stringLiteral = StringLiteralExprSyntax(first) {
+      hasStringLiterals = true
+      longestPounds = stringLiteral.openingPounds
+    }
+
+    while let concat = iter.next(), let stringComponent = iter.next() {
+      guard let concat = BinaryOperatorExprSyntax(concat), case .binaryOperator("+") = concat.operator.tokenKind else {
```

Comment:
A little nitpicky but `TokenKind` is equatable, so we don’t need to use `case` here, I think. I just think that `==` reads nicer. Also, we generally prefer to use `as` for downcasting.

```suggestion
      guard let concat = concat.as(BinaryOperatorExprSyntax.self), concat.operator.tokenKind == .binaryOperator("+") else {
```